examples/plugins/engine/ExampleWritePlugin.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Distributed under the OSI-approved Apache License, Version 2.0.  See
+ * accompanying file Copyright.txt for details.
+ *
+ * ExampleWritePlugin.h This plugin writes API calls out to a
+ * log file, as well as uses the BP4Writer for performing actual writes of
+ * data.
+ *
+ *  Created on: Jul 5, 2021
+ *      Author: Chuck Atkins <chuck.atkins@kitware.com>
+ *              Caitlin Ross <caitlin.ross@kitware.com>
+ */
+
+#ifndef EXAMPLEENGINEPLUGIN_H_
+#define EXAMPLEENGINEPLUGIN_H_
+
+#include <fstream>
+#include <memory>
+#include <string>
+
+#include "adios2/common/ADIOSMacros.h"
+#include "adios2/common/ADIOSTypes.h"
+#include "adios2/core/IO.h"
+#include "adios2/engine/bp4/BP4Writer.h"
+#include "adios2/engine/plugin/PluginEngineInterface.h"
+#include "adios2/helper/adiosComm.h"
+
+namespace adios2
+{
+namespace core
+{
+namespace engine
+{
+
+/** An engine interface to be used by the plugin infrastructure */
+class ExampleWritePlugin : public PluginEngineInterface
+{
+public:
+    ExampleWritePlugin(IO &io, const std::string &name, const Mode openMode,
+                       helper::Comm comm);
+    virtual ~ExampleWritePlugin();
+
+    StepStatus BeginStep(StepMode mode,
+                         const float timeoutSeconds = -1.0) override;
+    size_t CurrentStep() const override;
+    void PerformPuts() override;
+    void EndStep() override;
+    void Flush(const int transportIndex = -1) override;
+
+protected:
+    void Init() override;
+
+#define declare(T)                                                             \
+    void DoPutSync(Variable<T> &variable, const T *values) override;           \
+    void DoPutDeferred(Variable<T> &variable, const T *values) override;
+    ADIOS2_FOREACH_STDTYPE_1ARG(declare)
+#undef declare
+
+    void DoClose(const int transportIndex = -1) override;
+
+private:
+    std::ofstream m_Log;
+    std::unique_ptr<BP4Writer> m_Writer;
```

Comment:
Just curious. Is this assuming that all potential plugin engines will use BP4? What if a plugin engine wants to write data in its own way?

Reply:
No, this is just an example plugin, there's no reliance on any specific engines by the actual PluginEngine (as in the piece that loads/runs the plugin code). But this example is being updated to actually do its own work instead of using BP under the covers.